File: dynamic_programming/knapsack.cpp
#include <algorithm>
#include <iostream>
#include <set>
#include <vector>

// Structure to hold both optimal value and subset
struct Solution {
  int maxValue;
  std::set<int> selectedItems;
};

// Reconstruct the solution
void constructSolution(const std::vector<std::vector<int>> &dp,
                       const std::vector<int> &wt, int i, int j,
                       std::set<int> &optimalSet) {
  if (i > 0 && j > 0) {
    if (dp[i - 1][j] == dp[i][j]) {
      constructSolution(dp, wt, i - 1, j, optimalSet);
    } else {
      optimalSet.insert(
          i); // Using 1-based indexing for items as in Python example
      constructSolution(dp, wt, i - 1, j - wt[i - 1], optimalSet);
    }
  }
}

// 0/1 Knapsack
Solution knapsack(int W, const std::vector<int> &wt,
                  const std::vector<int> &val) {
  int n = wt.size();
  std::vector<std::vector<int>> dp(n + 1, std::vector<int>(W + 1, 0));

  for (int i = 1; i <= n; ++i) {
    for (int w = 1; w <= W; ++w) {
      if (wt[i - 1] <= w) {
        dp[i][w] =
            std::max(val[i - 1] + dp[i - 1][w - wt[i - 1]], dp[i - 1][w]);
      } else {
        dp[i][w] = dp[i - 1][w];
      }
    }
  }

  Solution sol;
  sol.maxValue = dp[n][W];
  constructSolution(dp, wt, n, W, sol.selectedItems);
  return sol;
}

int main() {
  std::vector<int> val = {3, 2, 4, 4};
  std::vector<int> wt = {4, 3, 2, 3};
  int W = 6;

  Solution sol = knapsack(W, wt, val);

  std::cout << "Optimal Value: " << sol.maxValue << std::endl;
  std::cout << "Selected Items: ";
  for (int item : sol.selectedItems) {
    std::cout << item << " ";
  }
  std::cout << std::endl;

  return 0;
}

File: dynamic_programming/knapsack.rs
use std::collections::HashSet;
use std::cmp;

pub fn knapsack(w_capacity: usize, wt: &[usize], val: &[usize]) -> (usize, HashSet<usize>) {
    let n = wt.len();
    let mut dp = vec![vec![0; w_capacity + 1]; n + 1];

    for i in 1..=n {
        for w in 1..=w_capacity {
            if wt[i - 1] <= w {
                dp[i][w] = cmp::max(
                    val[i - 1] + dp[i - 1][w - wt[i - 1]],
                    dp[i - 1][w],
                );
            } else {
                dp[i][w] = dp[i - 1][w];
            }
        }
    }

    let mut selected_items = HashSet::new();
    let mut temp_w = w_capacity;
    for i in (1..=n).rev() {
        if dp[i][temp_w] != dp[i - 1][temp_w] {
            selected_items.insert(i);
            temp_w -= wt[i - 1];
        }
    }

    (dp[n][w_capacity], selected_items)
}

fn main() {
    let val = vec![3, 2, 4, 4];
    let wt = vec![4, 3, 2, 3];
    let w = 6;

    let (max_value, selected_items) = knapsack(w, &wt, &val);
    println!("Optimal Value: {}", max_value);
    println!("Selected Items: {:?}", selected_items);
}
